File: Sources/Mobidex/Services/AcpClient.swift
import Foundation
import NIOCore

/// Thin ACP client for driving spec-compliant stdio agents (`grok agent stdio`,
/// `bunx @zed-industries/claude-code-acp`, ...) over a raw line transport on iOS.
///
/// Reuses `CodexLineTransport` (SSHRawExecTransport via `sshService.openRawExec` +
/// `SharedKMPBridge.acpShellCommand`).
///
/// `session/update` chunks are classified via bridged `AcpProtocolCore` and mapped (via
/// `acpClassificationToSessionItems`) into the exact `CodexThreadItem` model (reasoning, agentMessage,
/// toolCall, plan, agentEvent) already used by the conversation UI, so ACP responses render in
/// `ConversationView` / `ConversationSection` with zero new UI code.
///
/// Codex app-server path and all Codex launch/WS code remain 100% untouched.
actor AcpClient {
    nonisolated let events: AsyncStream<CodexAppServerEvent>
    nonisolated let sessionItems: AsyncStream<CodexThreadItem>  // The primary UI translation surface (mapper output)

    private let transport: CodexLineTransport
    private let rpcCore = SharedKMPBridge.makeRPCClientCore()
    private let acpCore = SharedKMPBridge.makeAcpProtocolCore()
    private let requestTimeoutSeconds: Int
    private let requestTimeoutNanoseconds: UInt64
    private var pending: [Int: CheckedContinuation<JSONValue, Error>] = [:]
    private var pendingTimeouts: [Int: Task<Void, Never>] = [:]
    private var isClosed = false
    private var readTask: Task<Void, Never>?
    private var currentSessionId: String?
    private var authMethodIds: [String] = []

    private let eventContinuation: AsyncStream<CodexAppServerEvent>.Continuation
    private let itemContinuation: AsyncStream<CodexThreadItem>.Continuation

    // Generous default: a cold `bunx @zed-industries/claude-code-acp` may download the package
    // on the host before answering initialize.
    init(transport: CodexLineTransport, requestTimeoutSeconds: Double = 120) {
        self.transport = transport
        self.requestTimeoutSeconds = max(1, Int(ceil(requestTimeoutSeconds)))
        self.requestTimeoutNanoseconds = UInt64(self.requestTimeoutSeconds) * 1_000_000_000

        let evStream = AsyncStream<CodexAppServerEvent>.makeStream()
        events = evStream.stream
        eventContinuation = evStream.continuation

        let itemStream = AsyncStream<CodexThreadItem>.makeStream()
        sessionItems = itemStream.stream
        itemContinuation = itemStream.continuation
    }

    deinit {
        readTask?.cancel()
    }

    func initialize() async throws {
        guard !isClosed else { throw CodexAppServerClientError.disconnected }
        let result = try await request(
            method: "initialize",
            params: SharedKMPBridge.acpInitializeParams()
        )
        // Remember the advertised auth methods for the auth_required retry in createSession.
        // (No "initialized" notification: ACP has none; spec agents log a decode error for it.)
        authMethodIds = SharedKMPBridge.acpAuthMethodIds(initializeResult: result)
    }

    /// Creates a new ACP session. When the agent answers auth_required (some agents, e.g. grok,
    /// require an explicit `authenticate` even when logged in on the host), authenticates with the
    /// first advertised method and retries once.
    func createSession(cwd: String, title: String? = nil) async throws -> String {
        guard !isClosed else { throw CodexAppServerClientError.disconnected }
        let params = SharedKMPBridge.acpSessionNewParams(cwd: cwd, title: title)
        let result: JSONValue
        do {
            result = try await request(method: "session/new", params: params)
        } catch let error as CodexAppServerClientError {
            guard case .appServer(let info) = error,
                  info.code == SharedKMPBridge.acpAuthRequiredErrorCode,
                  let methodId = authMethodIds.first
            else { throw error }
            _ = try await request(method: "authenticate", params: SharedKMPBridge.acpAuthenticateParams(methodId: methodId))
            result = try await request(method: "session/new", params: params)
        }
        // ACP typically returns { "sessionId": "..." } or the id directly; be tolerant
        let sid: String = {
            if case .object(let obj) = result, case .string(let s)? = obj["sessionId"] { return s }
            if case .string(let s) = result { return s }
            return "acp-\(UUID().uuidString.prefix(8))"
        }()
        currentSessionId = sid
        return sid
    }

    func sendPrompt(sessionId: String, text: String) async throws {
        guard !isClosed else { throw CodexAppServerClientError.disconnected }
        // Fire-and-forget prompt; streaming chunks arrive as session/update notifications
        let req = SharedKMPBridge.nextRequestLine(
            core: rpcCore,
            method: "session/prompt",
            params: SharedKMPBridge.acpSessionPromptParams(sessionId: sessionId, prompt: text)
        )
        try await transport.sendLine(req.line)
        // Do not await a result for the prompt itself in basic ACP; updates come asynchronously
    }

    /// Spec cancellation: `session/cancel` notification (the agent answers the prompt with stopReason=cancelled).
    func cancel(sessionId: String) async throws {
        guard !isClosed else { throw CodexAppServerClientError.disconnected }
        try await sendNotification(method: "session/cancel", params: SharedKMPBridge.acpSessionCancelParams(sessionId: sessionId))
    }

    /// Answers an agent -> client request (e.g. a `session/request_permission` outcome).
    func respondToServerRequest(id: JSONValue, result: JSONValue) async throws {
        try ensureOpenAndStartReadLoop()
        do {
            try await transport.sendLine(SharedKMPBridge.resultLine(core: rpcCore, id: id, result: result))
        } catch {
            let clientError = clientFacingError(error)
            await disconnect(error: clientError, message: clientError.localizedDescription, notify: true)
            throw clientError
        }
    }

    func close() async {
        if isClosed { return }
        isClosed = true
        readTask?.cancel()
        readTask = nil
        await transport.close()
        failPending(CodexAppServerClientError.disconnected)
        eventContinuation.finish()
        itemContinuation.finish()
    }

    // MARK: - Internal request / notification / read loop (modeled exactly on CodexAppServerClient)

    private func ensureOpenAndStartReadLoop() throws {
        guard !isClosed else {
            throw CodexAppServerClientError.disconnected
        }
        startReadLoopIfNeeded()
    }

    private func startReadLoopIfNeeded() {
        guard readTask == nil else { return }
        readTask = Task { [weak self] in
            await self?.readLoop()
        }
    }

    private func request(method: String, params: JSONValue?) async throws -> JSONValue {
        try ensureOpenAndStartReadLoop()
        let request = SharedKMPBridge.nextRequestLine(core: rpcCore, method: method, params: params)

        return try await withTaskCancellationHandler {
            try await withCheckedThrowingContinuation { continuation in
                pending[request.id] = continuation
                pendingTimeouts[request.id] = Task { [requestID = request.id, method, timeout = requestTimeoutNanoseconds] in
                    do { try await Task.sleep(nanoseconds: timeout) } catch { return }
                    self.timeoutRequest(id: requestID, method: method)
                }
                Task {
                    do {
                        try await transport.sendLine(request.line)
                    } catch {
                        let clientError = clientFacingError(error)
                        self.resolve(id: request.id, result: .failure(clientError))
                        Task { await self.disconnect(error: clientError, message: clientError.localizedDescription, notify: true) }
                    }
                }
            }
        } onCancel: {
            Task { await self.cancelRequest(id: request.id) }
        }
    }

    private func sendNotification(method: String, params: JSONValue?) async throws {
        try ensureOpenAndStartReadLoop()
        do {
            try await transport.sendLine(SharedKMPBridge.notificationLine(core: rpcCore, method: method, params: params))
        } catch {
            let clientError = clientFacingError(error)
            await disconnect(error: clientError, message: clientError.localizedDescription, notify: true)
            throw clientError
        }
    }

    private func readLoop() async {
        do {
            for try await line in transport.inboundLines {
                guard let data = line.data(using: .utf8), !data.isEmpty else { continue }
                let envelope: CodexRPCInboundEnvelope
                do {
                    envelope = try JSONDecoder().decode(CodexRPCInboundEnvelope.self, from: data)
                } catch {
                    throw CodexAppServerClientError.messageDecodeFailed(
                        "\(DecodeFailureFormatter.describe(error)). Line: \(DecodeFailureFormatter.preview(line))"
                    )
                }

                switch SharedKMPBridge.acpClassifyInbound(core: acpCore, envelope: (
                    id: envelope.id,
                    method: envelope.method,
                    params: envelope.params,
                    result: envelope.result,
                    error: envelope.error.map { CodexRPCErrorInfo(code: Int($0.code), message: $0.message) }
                )) {
                case .errorResponse(let id, let error):
                    let readable = CodexRPCErrorInfo(
                        code: error.code,
                        message: SharedKMPBridge.acpReadableError(code: error.code, message: error.message)
                    )
                    resolve(id: id, result: .failure(CodexAppServerClientError.appServer(readable)))
                case .resultResponse(let id, let result):
                    resolve(id: id, result: .success(result))
                case .serverRequest(let id, let method, let params):
                    eventContinuation.yield(.serverRequest(id: id, method: method, params: params))
                case .notification(let method, let params):
                    eventContinuation.yield(.notification(method: method, params: params))
                case .sessionUpdate(let classification):
                    let items = SharedKMPBridge.acpClassificationToSessionItems(classification)
                    for item in items {
                        itemContinuation.yield(item)
                    }
                case nil:
                    continue
                }
            }
            await disconnect(error: CodexAppServerClientError.disconnected, message: "ACP transport ended.", notify: true)
        } catch {
            let clientError = clientFacingError(error)
            await disconnect(error: clientError, message: clientError.localizedDescription, notify: true)
        }
    }

    private func disconnect(error: Error, message: String, notify: Bool) async {
        guard !isClosed else { return }
        isClosed = true
        readTask = nil
        await transport.close()
        failPending(error)
        if notify {
            eventContinuation.yield(.disconnected(message))
        }
        eventContinuation.finish()
        itemContinuation.finish()
    }

    private func resolve(id: Int, result: Result<JSONValue, Error>) {
        guard let continuation = pending.removeValue(forKey: id) else { return }
        pendingTimeouts.removeValue(forKey: id)?.cancel()
        continuation.resume(with: result)
    }

    private func timeoutRequest(id: Int, method: String) {
        resolve(id: id, result: .failure(CodexAppServerClientError.requestTimedOut(method: method, seconds: requestTimeoutSeconds)))
    }

    private func cancelRequest(id: Int) {
        resolve(id: id, result: .failure(CancellationError()))
    }

    private func failPending(_ error: Error) {
        let continuations = pending.values
        pending.removeAll()
        let timeouts = pendingTimeouts.values
        pendingTimeouts.removeAll()
        for t in timeouts { t.cancel() }
        for c in continuations { c.resume(throwing: error) }
    }

    private func clientFacingError(_ error: Error) -> Error {
        if error is CodexAppServerClientError { return error }
        if let ch = error as? ChannelError {
            return CodexAppServerClientError.transportClosed("ACP transport: \(ch)")
        }
        return error
    }
}

// MARK: - Debug conveniences (smoke / unit tests)
#if DEBUG
extension AcpClient {
    static func makeStub(transport: CodexLineTransport) -> AcpClient {
        AcpClient(transport: transport)
    }
}
#endif

File: android-app/src/main/java/mobidex/android/service/AcpClient.kt
package mobidex.android.service

import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.NonCancellable
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.TimeoutCancellationException
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.withContext
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.receiveAsFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withTimeout
import kotlinx.serialization.json.JsonElement
import java.util.concurrent.atomic.AtomicLong
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import mobidex.shared.AcpProtocolCore
import mobidex.shared.AcpRpcInboundEnvelope
import mobidex.shared.AcpRpcRequests
import mobidex.shared.CodexSessionItem
import mobidex.shared.JsonValue
import mobidex.shared.AcpSessionModels
import mobidex.shared.acpAuthMethodIds
import mobidex.shared.acpSessionModels
import mobidex.shared.encodeJsonLine
import mobidex.shared.toCodexSessionItems

/** An agent -> client JSON-RPC request (e.g. `session/request_permission`) awaiting a response. */
data class AcpServerRequest(
    val id: JsonValue,
    val method: String,
    val params: JsonValue?,
)

/** A created ACP session: its id plus the model state the agent advertised (null = no switching). */
data class AcpSession(
    val sessionId: String,
    val models: AcpSessionModels?,
)

/** JSON-RPC error from the agent, with the code preserved for auth_required handling. */
class AcpRpcException(val code: Int, message: String) : Exception(message)

/**
 * Thin ACP client for driving spec-compliant stdio agents (`grok agent stdio`,
 * `bunx @zed-industries/claude-code-acp`, ...) over a raw line transport.
 *
 * - Uses CodexLineTransport (obtained via SshService.openRawExec + RemoteAcpCommand.shellCommand)
 * - Leverages shared AcpProtocolCore for request building, inbound classification, and the
 *   chunk-to-CodexSessionItem mapper (so agent thoughts render as Reasoning, messages as
 *   AgentMessage, tools/plans via the existing ConversationSection UI with zero duplication).
 * - Authenticates on demand: when session/new fails with auth_required, retries once after
 *   `authenticate` with the first method the agent advertised (grok requires this even when
 *   logged in on the host).
 * - Surfaces agent -> client requests (permission prompts) on [serverRequests] for the
 *   ViewModel to answer via [respondToServerRequest].
 */
class AcpClient(
    private val transport: CodexLineTransport,
) {
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val itemsChannel = Channel<CodexSessionItem>(Channel.BUFFERED)
    private val serverRequestChannel = Channel<AcpServerRequest>(Channel.BUFFERED)
    private val pending = mutableMapOf<Long, CompletableDeferred<JsonElement>>()
    private val pendingMutex = Mutex()

    @Volatile
    private var closed = false
    private var currentSessionId: String? = null
    private val idCounter = AtomicLong(1)
    private var authMethodIds: List<String> = emptyList()

    private companion object {
        // Generous: a cold `bunx @zed-industries/claude-code-acp` may download the package
        // on the host before answering initialize.
        const val REQUEST_TIMEOUT_MS = 120_000L
    }

    /** Emits mapped CodexSessionItem instances (AgentMessage, Reasoning for thoughts, ToolCall, Plan, ...). */
    val sessionItems: Flow<CodexSessionItem> = itemsChannel.receiveAsFlow()

    /** Emits agent -> client requests (permission prompts) that must be answered via [respondToServerRequest]. */
    val serverRequests: Flow<AcpServerRequest> = serverRequestChannel.receiveAsFlow()

    // replay = 1: a disconnect that fires before the ViewModel's collector subscribes
    // (e.g. the transport dies right after createSession) must still be observed.
    private val _disconnects = MutableSharedFlow<String>(replay = 1)

    /** Emits once when the transport ends or fails after a successful start (parity with the iOS events stream). */
    val disconnects: SharedFlow<String> = _disconnects.asSharedFlow()

    init {
        scope.launch { readLoop() }
    }

    suspend fun initialize() {
        check(!closed) { "ACP client is closed." }
        val result = sendRequestAndAwait(AcpRpcRequests.initialize(id = nextId()))
        authMethodIds = acpAuthMethodIds(result.toSharedJsonValue())
    }

    /**
     * Creates a new ACP session and returns its id plus advertised model state. When the agent
     * answers auth_required, authenticates with its first advertised method and retries once.
     */
    suspend fun createSession(cwd: String, title: String? = null): AcpSession {
        check(!closed) { "ACP client is closed." }
        val result = try {
            sendRequestAndAwait(AcpRpcRequests.sessionNew(id = nextId(), cwd = cwd, title = title))
        } catch (error: AcpRpcException) {
            val methodId = authMethodIds.firstOrNull()
            if (error.code != AcpProtocolCore.AUTH_REQUIRED_ERROR_CODE || methodId == null) throw error
            sendRequestAndAwait(AcpRpcRequests.authenticate(id = nextId(), methodId = methodId))
            sendRequestAndAwait(AcpRpcRequests.sessionNew(id = nextId(), cwd = cwd, title = title))
        }
        val sid = extractSessionId(result) ?: error("session/new did not return a sessionId in result: $result")
        currentSessionId = sid
        return AcpSession(sessionId = sid, models = acpSessionModels(result.toSharedJsonValue()))
    }

    /** Switches the session's model to one of the ids advertised by [createSession]. */
    suspend fun setModel(sessionId: String, modelId: String) {
        check(!closed) { "ACP client is closed." }
        sendRequestAndAwait(AcpRpcRequests.sessionSetModel(id = nextId(), sessionId = sessionId, modelId = modelId))
    }

    /** Fire-and-forget: the prompt result (stopReason) only arrives when the whole turn ends. */
    suspend fun sendPrompt(sessionId: String, text: String) {
        check(!closed) { "ACP client is closed." }
        val req = AcpRpcRequests.sessionPrompt(id = nextId(), sessionId = sessionId, prompt = text)
        transport.sendLine(req.encodeJsonLine())
    }

    /** Spec cancellation: `session/cancel` notification (the agent answers the prompt with stopReason=cancelled). */
    suspend fun cancel(sessionId: String) {
        check(!closed) { "ACP client is closed." }
        transport.sendLine(AcpProtocolCore.notificationLine("session/cancel", AcpRpcRequests.sessionCancelParams(sessionId)))
    }

    /** Answers an agent -> client request (e.g. a permission prompt outcome). */
    suspend fun respondToServerRequest(id: JsonValue, result: JsonValue) {
        check(!closed) { "ACP client is closed." }
        transport.sendLine(AcpProtocolCore.resultLine(id, result))
    }

    suspend fun close() {
        if (!closed) {
            closed = true
            failAllPending(IllegalStateException("ACP client closed."))
            transport.close()
        }
        // Always close the channels (idempotent): a close() racing disconnect() must not
        // leave the reader's final send parked on a full buffer forever.
        itemsChannel.close()
        serverRequestChannel.close()
    }

    private fun nextId(): Long = idCounter.getAndIncrement() // monotonic to avoid collisions on rapid initialize + session/new (codex review P2)

    private suspend fun sendRequestAndAwait(req: mobidex.shared.CodexRpcRequest): JsonElement {
        val waiter = CompletableDeferred<JsonElement>()
        pendingMutex.withLock {
            // Re-check under the same lock failAllPending uses: registering after the fail
            // sweep would otherwise orphan the waiter until the timeout.
            check(!closed) { "ACP client is closed." }
            pending[req.id] = waiter
        }
        try {
            transport.sendLine(req.encodeJsonLine())
        } catch (e: Throwable) {
            // NonCancellable: if sendLine failed via cancellation, the suspending lock would
            // otherwise rethrow immediately and leak the pending entry.
            withContext(NonCancellable) { pendingMutex.withLock { pending.remove(req.id) } }
            waiter.completeExceptionally(e)
            throw e
        }
        try {
            return withTimeout(REQUEST_TIMEOUT_MS) { waiter.await() }
        } catch (e: TimeoutCancellationException) {
            pendingMutex.withLock { pending.remove(req.id) }
            throw IllegalStateException("ACP request ${req.method} timed out after ${REQUEST_TIMEOUT_MS / 1000}s.", e)
        } catch (e: CancellationException) {
            // Caller cancelled: drop the registration (a cancelled coroutine cannot take a
            // suspending lock, so do it under NonCancellable).
            withContext(NonCancellable) { pendingMutex.withLock { pending.remove(req.id) } }
            throw e
        }
    }

    private suspend fun readLoop() {
        try {
            transport.inboundLines.collect { line ->
                if (line.isBlank()) return@collect
                val message = AppJson.parseToJsonElement(line).jsonObject
                val envelope = message.toAcpEnvelope()
                val classification = AcpProtocolCore.classifyInbound(envelope) ?: return@collect

                when (classification.kind) {
                    "errorResponse" -> {
                        val id = classification.numericId ?: return@collect
                        val code = classification.error?.code ?: 0
                        val err = AcpProtocolCore.readableError(code, classification.error?.message ?: "ACP error")
                        pendingMutex.withLock { pending.remove(id) }?.completeExceptionally(AcpRpcException(code, err))
                    }
                    "resultResponse" -> {
                        val id = classification.numericId ?: return@collect
                        val resultEl = classification.result?.toJsonElement() ?: return@collect
                        pendingMutex.withLock { pending.remove(id) }?.complete(resultEl)
                    }
                    "sessionUpdate" -> {
                        // The key path: agent chunks → existing UI model via the shared mapper.
                        // Suspending send: backpressure the transport instead of silently
                        // dropping chat content when the consumer falls behind.
                        classification.toCodexSessionItems().forEach { item ->
                            itemsChannel.send(item)
                        }
                    }
                    "serverRequest" -> {
                        val id = classification.id ?: return@collect
                        val method = classification.method ?: return@collect
                        serverRequestChannel.send(AcpServerRequest(id = id, method = method, params = classification.params))
                    }
                    else -> {
                        // Ignore other notifications (vendor extensions, stopReason-less acks, ...).
                    }
                }
            }
            if (!closed) disconnect("ACP transport stream ended.")
        } catch (error: Throwable) {
            if (!closed) disconnect(error.message ?: "ACP read error: ${error::class.simpleName}")
        }
    }

    private suspend fun disconnect(message: String) {
        closed = true
        failAllPending(IllegalStateException(message))
        _disconnects.tryEmit(message)
        runCatching {
            itemsChannel.send(
                CodexSessionItem.AgentEvent(
                    id = "acp-disconnect-${System.currentTimeMillis()}",
                    label = "disconnected",
                    status = "error",
                    detail = message,
                )
            )
        }
        itemsChannel.close()
        serverRequestChannel.close()
    }

    private suspend fun failAllPending(error: Throwable) {
        val toFail = pendingMutex.withLock {
            val snapshot = pending.toMap()
            pending.clear()
            snapshot
        }
        toFail.values.forEach { it.completeExceptionally(error) }
    }

    private fun extractSessionId(result: JsonElement): String? {
        val obj = result as? JsonObject ?: return null
        return obj["sessionId"]?.jsonPrimitive?.contentOrNull
            ?: obj["session_id"]?.jsonPrimitive?.contentOrNull
            ?: obj["id"]?.jsonPrimitive?.contentOrNull
            ?: obj["session"]?.jsonObject?.get("id")?.jsonPrimitive?.contentOrNull
    }

    private fun JsonObject.toAcpEnvelope(): AcpRpcInboundEnvelope =
        AcpRpcInboundEnvelope(
            id = this["id"]?.toSharedJsonValue(),
            method = this["method"]?.jsonPrimitive?.contentOrNull,
            params = this["params"]?.toSharedJsonValue(),
            result = this["result"]?.toSharedJsonValue(),
            error = this["error"]?.jsonObject?.let { err ->
                mobidex.shared.CodexRpcErrorInfo(
                    code = err["code"]?.jsonPrimitive?.contentOrNull?.toIntOrNull() ?: 0,
                    message = err["message"]?.jsonPrimitive?.contentOrNull ?: "ACP error",
                )
            },
        )
}
